File: src/backend/common/logger/go/logger.go
package logger

import (
	"os"

	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
)

/*
Пакет logger предоставляет единый интерфейс логирования для всех Go-микросервисов.

Все логи выводятся в stdout в JSON-формате для централизованного сбора через Grafana Loki.
Формат логов соответствует единому стандарту TensorTalks для удобного парсинга и анализа.

Использование:
	logger := logger.New("auth-service", "1.0.0")
	logger.Info("service started", zap.String("port", "8081"))
	logger.Error("failed to connect", zap.Error(err))
*/

// Logger обёртка над zap.Logger с единым форматом для всех микросервисов.
type Logger struct {
	*zap.Logger
	serviceName string
	version     string
}

// Config конфигурация логгера.
type Config struct {
	ServiceName string // Имя сервиса (например, "auth-service")
	Version     string // Версия сервиса (например, "1.0.0")
	Level       string // Уровень логирования: "debug", "info", "warn", "error" (по умолчанию "info")
	Environment string // Окружение: "development", "production" (по умолчанию "production")
}

// New создаёт новый экземпляр логгера с единым форматом для микросервисов.
// Все логи выводятся в stdout в JSON-формате для сбора через Loki.
func New(serviceName, version string) *Logger {
	return NewWithConfig(Config{
		ServiceName: serviceName,
		Version:     version,
		Level:       "info",
		Environment: "production",
	})
}

// NewWithConfig создаёт логгер с кастомной конфигурацией.
func NewWithConfig(cfg Config) *Logger {
	// Определяем уровень логирования
	var level zapcore.Level
	switch cfg.Level {
	case "debug":
		level = zapcore.DebugLevel
	case "warn":
		level = zapcore.WarnLevel
	case "error":
		level = zapcore.ErrorLevel
	default:
		level = zapcore.InfoLevel
	}

	// Настройка encoder config для JSON-формата
	encoderConfig := zapcore.EncoderConfig{
		TimeKey:        "timestamp",
		LevelKey:       "level",
		NameKey:        "logger",
		CallerKey:      "caller",
		FunctionKey:    zapcore.OmitKey,
		MessageKey:     "message",
		StacktraceKey:  "stacktrace",
		LineEnding:     zapcore.DefaultLineEnding,
		EncodeLevel:    zapcore.LowercaseLevelEncoder,
		EncodeTime:     zapcore.ISO8601TimeEncoder,
		EncodeDuration: zapcore.SecondsDurationEncoder,
		EncodeCaller:   zapcore.ShortCallerEncoder,
	}

	// В development режиме используем более читаемый формат
	if cfg.Environment == "development" {
		encoderConfig.EncodeLevel = zapcore.CapitalColorLevelEncoder
		encoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
	}

	// Создаём core для вывода в stdout
	core := zapcore.NewCore(
		zapcore.NewJSONEncoder(encoderConfig),
		zapcore.AddSync(os.Stdout),
		zap.NewAtomicLevelAt(level),
	)

	// Создаём logger с полями сервиса
	logger := zap.New(core, zap.AddCaller(), zap.AddStacktrace(zapcore.ErrorLevel))

	// Добавляем стандартные поля для всех логов
	logger = logger.With(
		zap.String("service", cfg.ServiceName),
		zap.String("version", cfg.Version),
		zap.String("environment", cfg.Environment),
	)

	return &Logger{
		Logger:      logger,
		serviceName: cfg.ServiceName,
		version:     cfg.Version,
	}
}

// WithRequestID добавляет request_id к логгеру для трейсинга запросов.
func (l *Logger) WithRequestID(requestID string) *Logger {
	return &Logger{
		Logger:      l.Logger.With(zap.String("request_id", requestID)),
		serviceName: l.serviceName,
		version:     l.version,
	}
}

// WithUserID добавляет user_id к логгеру.
func (l *Logger) WithUserID(userID string) *Logger {
	return &Logger{
		Logger:      l.Logger.With(zap.String("user_id", userID)),
		serviceName: l.serviceName,
		version:     l.version,
	}
}

// WithFields добавляет произвольные поля к логгеру.
func (l *Logger) WithFields(fields ...zap.Field) *Logger {
	return &Logger{
		Logger:      l.Logger.With(fields...),
		serviceName: l.serviceName,
		version:     l.version,
	}
}

File: src/backend/common/logger/python/logger.py
"""
Единый модуль логирования для всех Python-микросервисов TensorTalks.

Все логи выводятся в stdout в JSON-формате для централизованного сбора через Grafana Loki.
Формат логов соответствует единому стандарту TensorTalks для удобного парсинга и анализа.

Использование:
    from logger import get_logger
    
    logger = get_logger("auth-service", "1.0.0")
    logger.info("service started", extra={"port": 8081})
    logger.error("failed to connect", exc_info=True)
"""

import json
import logging
import sys
from datetime import datetime
from typing import Any, Dict, Optional


class JSONFormatter(logging.Formatter):
    """Форматтер для вывода логов в JSON-формате для Loki."""

    def format(self, record: logging.LogRecord) -> str:
        """Форматирует запись лога в JSON."""
        log_data: Dict[str, Any] = {
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "level": record.levelname.lower(),
            "message": record.getMessage(),
            "service": getattr(record, "service", "unknown"),
            "version": getattr(record, "version", "unknown"),
            "environment": getattr(record, "environment", "production"),
        }

        # Добавляем caller информацию
        if record.pathname:
            log_data["caller"] = f"{record.pathname}:{record.lineno}"

        # Добавляем request_id если есть
        if hasattr(record, "request_id"):
            log_data["request_id"] = record.request_id

        # Добавляем user_id если есть
        if hasattr(record, "user_id"):
            log_data["user_id"] = record.user_id

        # Добавляем дополнительные поля из extra
        if hasattr(record, "extra_fields"):
            log_data.update(record.extra_fields)

        # Добавляем exception информацию если есть
        if record.exc_info:
            log_data["stacktrace"] = self.formatException(record.exc_info)

        return json.dumps(log_data, ensure_ascii=False)


def get_logger(
    service_name: str,
    version: str,
    level: str = "INFO",
    environment: str = "production",
) -> logging.Logger:
    """
    Создаёт и настраивает логгер для микросервиса.

    Args:
        service_name: Имя сервиса (например, "auth-service")
        version: Версия сервиса (например, "1.0.0")
        level: Уровень логирования (DEBUG, INFO, WARNING, ERROR, CRITICAL)
        environment: Окружение (development, production)

    Returns:
        Настроенный logger
    """
    logger = logging.getLogger(service_name)
    logger.setLevel(getattr(logging, level.upper(), logging.INFO))

    # Удаляем существующие handlers чтобы избежать дублирования
    logger.handlers.clear()

    # Создаём handler для stdout
    handler = logging.StreamHandler(sys.stdout)
    handler.setFormatter(JSONFormatter())

    logger.addHandler(handler)
    logger.propagate = False

    # Добавляем стандартные атрибуты к записям
    old_factory = logging.getLogRecordFactory()

    def record_factory(*args, **kwargs):
        record = old_factory(*args, **kwargs)
        record.service = service_name
        record.version = version
        record.environment = environment
        return record

    logging.setLogRecordFactory(record_factory)

    return logger


def with_request_id(logger: logging.Logger, request_id: str) -> logging.Logger:
    """
    Создаёт адаптер логгера с request_id для трейсинга запросов.

    Args:
        logger: Базовый логгер
        request_id: ID запроса

    Returns:
        Адаптер логгера
    """
    return logging.LoggerAdapter(logger, {"request_id": request_id})


def with_user_id(logger: logging.Logger, user_id: str) -> logging.Logger:
    """
    Создаёт адаптер логгера с user_id.

    Args:
        logger: Базовый логгер
        user_id: ID пользователя

    Returns:
        Адаптер логгера
    """
    return logging.LoggerAdapter(logger, {"user_id": user_id})


def with_fields(logger: logging.Logger, **fields: Any) -> logging.Logger:
    """
    Создаёт адаптер логгера с дополнительными полями.

    Args:
        logger: Базовый логгер
        **fields: Дополнительные поля для логирования

    Returns:
        Адаптер логгера
    """
    return logging.LoggerAdapter(logger, fields)
